src/compiler/checker.ts

```diff
@@ -38424,6 +38392,45 @@ namespace ts {
             }
         }
 
+        function checkUnmatchedJSDocParameters(node: SignatureDeclaration) {
+            const jsdocParameters = filter(getJSDocTags(node), isJSDocParameterTag);
+            if (length(jsdocParameters) === 0) return;
+
+            const isJs = isInJSFile(node);
+            const parameters = new Set<__String>();
+            const excludedParameters = new Set<number>();
+            forEach(node.parameters, ({ name }, index) => {
+                if (isIdentifier(name)) {
+                    parameters.add(name.escapedText);
+                }
+                if (isBindingPattern(name)) {
+                    excludedParameters.add(index);
+                }
+            });
+
+            const containsArguments = containsArgumentsReference(node);
+            if (containsArguments) {
+                const lastJSDocParam = lastOrUndefined(jsdocParameters);
+                if (lastJSDocParam && isIdentifier(lastJSDocParam.name) && lastJSDocParam.typeExpression &&
+                    lastJSDocParam.typeExpression.type && !parameters.has(lastJSDocParam.name.escapedText) && !isArrayType(getTypeFromTypeNode(lastJSDocParam.typeExpression.type))) {

```

Comment:
How does the `parameters.has` check relate to the old code? I can't derive it so I'm probably missing something.

Reply:
@sandersn It doesn't relate to the old code. The previous version calls `getParameterSymbolFromJSDoc`/`getHostSignatureFromJSDoc` for each parameter. The proposed version works the opposite. To get tags in TS/JS I moved this check to the `signature` level. Then, to compare parameters with tags, instead of using `indexOf` (in this case I think `indexOf` gives O(n) complexity) for each tag., I moved the parameters to `Set`. The `Set` is then used in the `jsdocParameters` loop to check for existing parameters by `JSDocParameter` name. Maybe this is the wrong direction, and a different approach should be taken here :(.